### PR TITLE
add Wrong Type work around to snmp_get function

### PR DESCRIPTION
**Pull request by peejaychilds**
Please give a short description what your pull request is for

A regression in JunOS means snmp_get function calls for Temperature fail to return data.   On discovery this means the sensor_limit_low is set to -10 and the sensor_limit is set to 20 (since sensor_current = 0).    This causes false alarms.

I note there already exists a work around for this in snmpwalk functions, but not snmpget function.

Adding the work around to snmpget function resolves the issue until we can the vendor fixes things.

**JunOS 22.2R3-S1.6**

```
snmpget -v2c -c COMMUNITY -Oqv -m JUNIPER-MIB -M /opt/librenms/mibs:/opt/librenms/mibs/junos udp:HOST-JUNOS-22-2R3:161 jnxOperatingTemp.9.1.0.0
Wrong Type (should be Gauge32 or Unsigned32): 35
```

```
'-D all'
(...)
trace: snmp_pdu_parse(): snmp_api.c, 4749:
dumph_recv:         Value
dumpx_recv:          02 01 23
dumpv_recv:            Integer:	35 (0x23)
(...)
```

**JunOS 20.4R3-S3.4**

```
snmpget -v2c -c COMMUNITY -Oqv -m JUNIPER-MIB -M /opt/librenms/mibs:/opt/librenms/mibs/junos udp:HOST-JUNOS-20-4R3:161 jnxOperatingTemp.7.1.0.0
52
```

```
'-D all'
(...)
trace: snmp_pdu_parse(): snmp_api.c, 4749:
dumph_recv:         Value
dumpx_recv:          42 01 34
dumpv_recv:            UInteger:	52 (0x34)
(...)
```

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [ ] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [ ] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Review by VVelox — CHANGES_REQUESTED** [review]
Okay. Sweet. This looks good.

That said it appears this results in the reordering of stuff for some of the tests, so it looks like those are going to need updated. Looking over the one that is failing, I don't see any data points that are going missing, just this resulting in them being ordered different.

**Comment by peejaychilds** [comment]
> Okay. Sweet. This looks good.
> 
> That said it appears this results in the reordering of stuff for some of the tests, so it looks like those are going to need updated. Looking over the one that is failing, I don't see any data points that are going missing, just this resulting in them being ordered different.

I couldn't figure out why the test results would change based on the very small change.   I figured it was something to do with the snmp test harness occasionally spitting out different things?

I'm not sure how to resolve it.

**Comment by VVelox** [comment]
> > Okay. Sweet. This looks good.
> > That said it appears this results in the reordering of stuff for some of the tests, so it looks like those are going to need updated. Looking over the one that is failing, I don't see any data points that are going missing, just this resulting in them being ordered different.
> 
> I couldn't figure out why the test results would change based on the very small change. I figured it was something to do with the snmp test harness occasionally spitting out different things?
> 
> I'm not sure how to resolve it.

I am assuming there is something funny there as well and it being polled differently now than before, but appears to be finding the same data.

That said I am far from the best person to ask for anything outside of app related stuff when it comes to pooling unfortunately.

**Comment by murrant** [comment]
Looks like it discovered an additional sensor. Likely, you fixed some other cases, but they will need to be double checked before updating test data.

Remember that those tests are just a snapshot of the data, right or wrong.  So we can detect changes.

File: includes/snmp.inc.php [hunk 1]
```diff
@@ -305,6 +305,10 @@ function snmp_get($device, $oid, $options = null, $mib = null, $mibdir = null)
     $measure->manager()->recordSnmp($measure->end());
     if (preg_match('/(No Such Instance|No Such Object|No more variables left|Authentication failure)/i', $data)) {
         return false;
+    } elseif (preg_match('/Wrong Type(.*)should be/', $data)) {
```
Comment (Jellyfrog) [inline includes/snmp.inc.php:308]: This should probably be a str_starts_with() instead

File: includes/snmp.inc.php [hunk 2]
```diff
@@ -305,6 +305,10 @@ function snmp_get($device, $oid, $options = null, $mib = null, $mibdir = null)
...
+        $data = preg_replace('/Wrong Type \(should be .*\): /', '', $data);
```
Comment (Jellyfrog) [inline includes/snmp.inc.php:309]: Super nitpick is to rewrite it in a slightly more efficient way, `/Wrong Type \(should be [^:]+: /`

File: includes/snmp.inc.php (hunk 1 above)
Comment (peejaychilds) [reply includes/snmp.inc.php:308]: I just copied https://github.com/librenms/librenms/blob/master/includes/snmp.inc.php#L424

File: includes/snmp.inc.php (hunk 2 above)
Comment (peejaychilds) [reply includes/snmp.inc.php:309]: I just copied https://github.com/librenms/librenms/blob/master/includes/snmp.inc.php#L424

Comment (peejaychilds) [reply includes/snmp.inc.php:309]: fwiw we have a few instances of 'handling' this

We have
 snmp_get
   https://github.com/librenms/librenms/blob/master/includes/snmp.inc.php#L302
 snmp_walk
   https://github.com/librenms/librenms/blob/master/includes/snmp.inc.php#L424
 snmp_get_multi_oid
   https://github.com/librenms/librenms/blob/master/includes/snmp.inc.php#L252

 tests/Unit/SnmpResponseTest.php
 testErrorHandling
   https://github.com/librenms/librenms/blob/master/tests/Unit/SnmpResponseTest.php#L280

 LibreNMS/Util/ModuleTestHelper.php
 convertSnmpToSnmprec
   https://github.com/librenms/librenms/blob/master/LibreNMS/Util/ModuleTestHelper.php#L374

 LibreNMS/Data/Source/SnmpResponse.php
   __construct
   https://github.com/librenms/librenms/blob/master/LibreNMS/Data/Source/SnmpResponse.php#L55

🤷‍♂️

Help?

Comment (murrant) [reply includes/snmp.inc.php:309]: Just convert everything to SnmpQuery :D

**Comment by murrant** [comment]
This seems questionable.  It might be easier to just convert snmp_get() to SnmpQuery::get()

**Comment by Jellyfrog** [comment]
I agree, the old functions are deprecated, better to move to to new way, which only requires changing in 1 place

**Comment by murrant** [comment]
Although, it seems this is using the yaml... so yeah will need a big test data update if it affects other output like this.  But at least it would have a smaller impact.

**Comment by murrant** [comment]
Does this fix it?  I was looking at things and it seems it should use snmpwalk_cache_oid()

**Comment by peejaychilds** [comment]
This probably doesn't impact the original case anymore (JunOS temperature sensor wrong type issue) since https://github.com/librenms/librenms/pull/15118

`librenms$git checkout 23.6.0
Note: switching to '23.6.0'.
(...)
HEAD is now at eaf025341 Bump version to 23.6.0
librenms$find . -type f -print0 | xargs -0 grep snmp_get\( | grep -i junos
./includes/polling/sensors/ber/junos.inc.php:    $sensor_exp_value = snmp_get($device, 'JNX-OPT-IF-EXT-MIB::jnxoptIfOTNPMCurrentFECBERExponent.' . $sensor['sensor_index'] . '.1', '-Oqv');
./includes/discovery/sensors/temperature/junose.inc.php:            $entPhysicalIndex = snmp_get($device, 'juniSystemTempPhysicalIndex.' . $index, '-Oqv', 'Juniper-System-MIB', 'junose');
./includes/discovery/sensors/temperature/junose.inc.php:            $descr = snmp_get($device, 'entPhysicalDescr.' . $entPhysicalIndex, '-Oqv', 'ENTITY-MIB');
./includes/discovery/sensors/temperature/junos.inc.php:        $descr = snmp_get($device, $descr_oid, '-Oqv', 'JUNIPER-MIB', 'junos');
./includes/discovery/sensors/temperature/junos.inc.php:        $temperature = snmp_get($device, $temperature_oid, '-Oqv', 'JUNIPER-MIB', 'junos');
./includes/discovery/vlans/junos.inc.php:$vlanversion = snmp_get($device, 'dot1qVlanVersionNumber.0', '-Oqv', 'IEEE8021-Q-BRIDGE-MIB');
librenms$git checkout master
Previous HEAD position was eaf025341 Bump version to 23.6.0
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
librenms$find . -type f -print0 | xargs -0 grep snmp_get\( | grep -i junos
./includes/discovery/sensors/temperature/junose.inc.php:            $entPhysicalIndex = snmp_get($device, 'juniSystemTempPhysicalIndex.' . $index, '-Oqv', 'Juniper-System-MIB', 'junose');
./includes/discovery/sensors/temperature/junose.inc.php:            $descr = snmp_get($device, 'entPhysicalDescr.' . $entPhysicalIndex, '-Oqv', 'ENTITY-MIB');
./includes/discovery/vlans/junos.inc.php:$vlanversion = snmp_get($device, 'dot1qVlanVersionNumber.0', '-Oqv', 'IEEE8021-Q-BRIDGE-MIB');`

So in we could close this, or stuff it in for feature parity between snmp_get and other legacy snmp retrieval functions ....

**Comment by murrant** [comment]
It can't be merged until tests succeed, so if you want to close it, go ahead.

**Comment by peejaychilds** [comment]
hmm... looks like some Huawei UPS 2000 tests fail ... I wonder if this is because the support for that device included calls to snmp_get with MIBs that had data types that did not match the return types from the snmprec/device, this patch fixes that, and now we have additional sensors showing up on this device.

I'm not sure I have enough current competency to either make that determination, or know how to resolve it.

**Comment by electrocret** [comment]
@peejaychilds , The failed tests read to me like Libre's now successfully picking up more sensors from the snmprec thanks to your PR, which is causing it to not match up with the json test output. 

Ie. There's nothing wrong with your work, you just need to regenerate the Test json from the snmprec.

**Comment by peejaychilds** [comment]
whoops did a git screw-the-pooch ...

**Pull request closed without merging**